### PR TITLE
Fix workflow for running Africa

### PR DESCRIPTION
**Pull request by davide-f**
# Closes # (if applicable).

## Changes proposed in this Pull Request
I openned this PR to describe current efforts in enabling the workflow to run for Africa.
This PR integrates #284 

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [ ] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by davide-f** [comment]
@Eddy-JV @hazemakhalek @energyLS this PR is ready for a first revision of the procedure and highlights.
But do not merge yet

**Comment by davide-f** [comment]
FYI this PR has run for Africa locally but the model is infeasible.
Numbers need checks

**Comment by davide-f** [comment]
@hazemakhalek unfortunately the use of 40000TWh of biomass potential didn't solve the issue

File: scripts/prepare_sector_network.py [hunk 1]
```diff
@@ -2032,20 +2037,21 @@ def add_services(n, costs):
         p_set=-co2,
     )
 
-    p_set_gas = (
-        profile_residential * energy_totals.loc[countries, "services gas"].sum() * 1e6
-    )
+    p_set_gas = 1e6 * profile_residential.mul(
+        energy_totals["services gas"], level=0
```
Comment (doneachh) [inline scripts/prepare_sector_network.py:2041]: What do you think: Does it make sense to write a method for setting p_set_gas, p_set_oil, p_set_biomass and p_set_elec as its often recurring?

something like
`def calculate_p_set(type):
        return 1e6 * profile_residential.mul(
        energy_totals[type], level=0
        ).droplevel(level=0, axis=1)
`

File: scripts/prepare_sector_network.py [hunk 2]
```diff
@@ -2059,23 +2065,25 @@ def add_services(n, costs):
...
 
     n.madd(
         "Load",
-        nodes,
+        spatial.nodes,
         suffix=" agriculture oil",
         bus=spatial.oil.nodes,
         carrier="agriculture oil",
-        p_set=nodal_energy_totals.loc[nodes, "agriculture oil"] * 1e6 / 8760,
+        p_set=nodal_energy_totals.loc[spatial.nodes, "agriculture oil"] * 1e6 / 8760,
```
Comment (doneachh) [inline scripts/prepare_sector_network.py:2083]: the same here: does it make sense to write a method for nodal_energy_totals.loc[spatial.nodes, "type"] * 1e6 / 8760?

**Review by doneachh — COMMENTED** [review]
Does make sense to me! Great work! :)

**Comment by doneachh** [comment]
@davide-f could you summarize for me, which tasks are open at this point for getting serious africa modeling results? 
- model doesnt have enough biomass potential
- Missing H2-Buses for GHA
- Something else as well?

**Comment by hazemakhalek** [comment]
> @hazemakhalek unfortunately the use of 40000TWh of biomass potential didn't solve the issue

Out of experience, 90% of the cases the problem is either biomass potential, oil potential or co2 emission limit

**Comment by hazemakhalek** [comment]
The PR looks alright to me, I am not able to run it though as I'm currently out of office. All the logic seem alright to me so if you tested it @davide-f and the CI passes I think we can merge it

**Comment by davide-f** [comment]
> The PR looks alright to me, I am not able to run it though as I'm currently out of office. All the logic seem alright to me so if you tested it @davide-f and the CI passes I think we can merge it

Great to hear! and many thanks :)

I'll wait merging it till at least [this is merged](https://github.com/pypsa-meets-earth/pypsa-earth/pull/989).
That leads to some little issues in the execution that can apply to -sec.
I'm locally testing it for Africa (-sec version) and World (only power).

**Comment by davide-f** [comment]
This has worked for Africa; Ready to review and finalize :)

**Comment by davide-f** [comment]
Many thanks! @doneachh !
One last update to be merged after review in pypsa-earth though:
https://github.com/pypsa-meets-earth/pypsa-earth/pull/992

**Comment by davide-f** [comment]
Moreover, this PR needs some checks: for MA it doesn't work well. I'm investigating

File: scripts/prepare_sector_network.py [hunk 3]
```diff
@@ -1168,9 +1173,9 @@ def add_industry(n, costs):
 
     # CARRIER = FOSSIL GAS
 
-    nodes = (
-        pop_layout.index
-    )  # TODO where to change country code? 2 letter country codes.
+    # nodes = (
+    #     pop_layout.index
+    # )  # TODO where to change country code? 2 letter country codes.
```
[the remark this answers is not in the document]
Comment (davide-f) [reply scripts/prepare_sector_network.py:1178]: Yeah, that is an old comment. Removing

File: scripts/prepare_sector_network.py (hunk 1 above)
Comment (davide-f) [reply scripts/prepare_sector_network.py:2041]: Agree; done.
However, in the future we may consider restyling this approach.
To me, it seems that similar operations are repeated for multiple entities; we could create a for loop and automatize the operations for those terms for example.

**Comment by davide-f** [comment]
> @davide-f could you summarize for me, which tasks are open at this point for getting serious africa modeling results?
> 
> * model doesnt have enough biomass potential
> * Missing H2-Buses for GHA
> * Something else as well?

Hello :D
Sorry my late reply, I didn't see your review but I've answered all your comments.

Regarding the rest:
- I've rerun the model with the latest changes, including both increased biomass (4e6 TWh/a), biogass (2e5 TWh/a) and oil (1e5 TWh) potentials --> the problem still is infeasible (or it has numerical issues)
- I confirm some missing buses for some gas pipelines that need some fixing. I'd recommend that we could improve the modelling in pypsa-earth so that when using the gadm clustering, each region shall have at least a bus. This should fix this problem

Overall, I feel like there is the need for some debugging for larger regions.
I'd recommend to start "easy": rather then modelling Africa as a whole, maybe start with a power pool, then follow to other power pools and finally the whole model. That is to get confidence of the results at regional level.
We need to check how the model performs with multiple countries: in this PR I aimed to fix the limitations I saw in prepare_sector_network that was coded to represent a single country, but there may be other limitations in the whole code. That needs some checking

**Comment by davide-f** [comment]
For me this PR is ready to go.
I'd recommend somebody to rerun it locally as well on MA to have a crosscheck.

**Comment by davide-f** [comment]
I think I discovered the issue and the problem relates to isolated nodes.
What I did is several testing. The latest one that helped me achieve this insight is:
1. run MA only with 20 nodes -> working
2. run MA+DZ with 20 nodes -> failing due to numerical issues
3. run MA+DZ with 15 nodes and varying CO2 limits -> failing
4. run DZ only with 10 nodes -> failing
5. checked the DZ-only network that was supposed to have only 10 nodes and instead it has 36

I see 2 options:
1. merge this PR
2. first rerun a MA scenario again comparing original and new approach and merge it

**Review by davide-f — COMMENTED** [review]
I'd propose to merge the PR as is and then revise it later; otherwise the workflow is deviating too much

**Comment by davide-f** [comment]
Tested for MA using upstream main and this PR.
the objective value changes slightly, about 1-2%, the reasons are the following:
1. this PR fixes a bug in the heat demand that was previously omitted
2. the update of the submodule has changes slightly something

If that's fine for you, we can merge

**Comment by doneachh** [comment]
@davide-f Fine for me!

**Comment by davide-f** [comment]
Africa has successfully executed locally :) merging
